pkg/resolution/resolver/git/git.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2022 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+)
+
+type repository struct {
+	url       string
+	directory string
+	revision  string
+}
+
+func resolveRepository(ctx context.Context, url, revision string) (*repository, func(), error) {
+	urlParts := strings.Split(url, "/")
+	repoName := urlParts[len(urlParts)-1]
+	tmpDir, err := os.MkdirTemp("", repoName+"-*")
+	if err != nil {
+		return nil, func() {}, err
+	}
+	cleanupFunc := func() {
+		os.RemoveAll(tmpDir)
+	}
+
+	repo := repository{
+		url:       url,
+		directory: tmpDir,
+		revision:  revision,
+	}
+
+	_, err = repo.execGit(ctx, "clone", repo.url, tmpDir, "--depth=1", "--no-checkout")
+	if err != nil {
+		return nil, cleanupFunc, err
+	}
+
+	_, err = repo.execGit(ctx, "fetch", "origin", repo.revision, "--depth=1")
+	if err != nil {
+		return nil, cleanupFunc, err
+	}
+
+	_, err = repo.execGit(ctx, "checkout", "FETCH_HEAD")
+	if err != nil {
+		return nil, cleanupFunc, err
+	}
+
+	revisionSha, err := repo.execGit(ctx, "rev-list", "-n1", "HEAD")
+	if err != nil {
+		return nil, cleanupFunc, err
+	}
+	repo.revision = strings.TrimSpace(string(revisionSha))
+
+	return &repo, cleanupFunc, nil
```

Comment:
I initially simply did a full `git clone` and then simply did `git checkout`. However doing a shallow clone with `--no-checkout`, `fetch`ing the revision, and then checking the revision out improved the time and space performance significantly. Using the same repository mentioned in [the original issue](https://git-scm.com/docs/protocol-capabilities#_allow_reachable_sha1_in_want) as a benchmark, using a shallow clone reduced the clone time from ~110 seconds by a factor of ten. Using `--no-checkout` similarly reduced the disk space and time slightly (before the necessary revision was checked-out)